doc/reference/trd-appid.md

```diff
@@ -0,0 +1,396 @@
+
+4. Credentials in Tock Binary Format Headers
+===============================
+
+To support credentials in Tock binaries, the Tock Binary Format has
+a `TbfHeaderV2Credentials` header. This header is variable length
+and has two fields:
+
+```rust
+pub struct TbfHeaderV2Credentials {
+    format: TbfHeaderV2CredentialsType,
+    data: &[u8],
+}  
+```
+
+The `TbfHeaderV2CredentialsType` defines the format and size of `data`
+field. A `TbfHeaderV2CredentialsType` value MUST have a fixed
+data size and format. Currently supported values are:
+
+```rust
+pub enum TbfHeaderV2CredentialsType {
+    CleartextID = 0,
+    Rsa3072Key = 1,
+    Rsa4096Key = 2,
+    Rsa3072KeyWithID = 3,
+    Rsa4096KeyWithID = 4,
+}
+```
+
+**These are not intended to be final or prescriptive. They are merely some examples
+of what kind of information we might put here. Among other things, the exact format 
+of the data blocks needs to be more precise. -pal**
+
+The `CleartextID` value has a data length of 8 bytes. It contains a 64-bit number in
+big-endian format representing an application identifier.
+
+The `Rsa3072Key` value has a data of length of 768 bytes. It contains a public 3072-bit
+RSA key (384 bytes), followed by a 384-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary, encrypted by the private key 
+of the public key in the header.
+
+The `Rsa4096Key` value has a data of length of 1024 bytes. It contains a public 4096-bit
+RSA key (512 bytes), followed by a 512-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary, encrypted by the private key 
+of the public key in the header.
+
+The `Rsa3072KeyWithID` value has a data of length of 768 bytes. It contains a public 3072-bit
+RSA key (384 bytes), followed by a 384-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary followed by a 32-bit application
+ID, encrypted by the private key of the public key in the header.
+
+The `Rsa4096KeyWithID` value has a data of length of 1024 bytes. It contains a public 4096-bit
+RSA key (512 bytes), followed by a 512-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary followed by a 32-bit application
+ID, encrypted by the private key of the public key in the header.
+
+
+4 `Verifier` trait
+===============================
+
+The `Verifier` trait defines an interface to a module that accepts,
+passes on, or rejects application credentials. When a Tock board
+asks the kernel to load processes, it passes a reference to a 
+`Verifier`, which the kernel uses to check credentials.
+
+
+```rust
+pub enum VerificationResult {
+  Accept,
+  Pass,
+  Reject
+}
+
+pub trait Verifier {
+  fn require_credential(&self) -> bool;
+  fn check_credentials(&self, 
+                       credentials: &TbfHeaderV2Credentials, 
+                       binary: &mut [u8]) -> VerificationResult;
```

Comment:
Don't `credentials` and `binary` alias here? We would want the TBF headers to be included in signatures.

There's also the complication that the signature cannot be included in the signed blob. I think the traditional solution is for the signature to be replaced with a string of zeroes during signing and verification, which requires `Verifier` to know where the credential header is within the binary.

Reply:
`binary` was intended to be the application binary, not process binary. As I wrote this I took note of the fact that this didn't cover the other headers. But I wasn't quite sure how we wanted to handle the headers (are *all* of them signed) so kind of punted to discussion.

`binary` should definitely be not `mut`. But yes, they'll need to alias. I'll think about a good API to handle the zero-ing. It might be you're passed two slices and a length of the intervening zeroes.

Reply:
There are fields in the TBF headers that must be covered by the signature (e.g. `init_offset`).

Signing some but not all of the headers would get very messy very fast. I hope we don't have a use case for cryptographically verified IDs that requires unsigned headers, but we need to confirm that.

Reply:
I agree -- all headers except for Credentials should be signed. Perhaps this means that Credentials need to be a contiguous block? Or a contiguous block at the end of the headers?

Reply:
I didn't even think about the fact there can be multiple Credentials headers. That certainly makes things more complicated.

In retrospect, I think I was confusing signatures and checksums when I suggested replacing the string with zeroes. I'm not sure if there's a standard practice for embedding a signature into a signed blob.

Reply:
This has been resolved by now having credentials as footers. The `binary` argument is no longer `mut`, and now covers only the part of the process binary which are covered by integrity; it therefore does not cover the credentials stored in footers.